client/protocol.go

```diff
@@ -0,0 +1,413 @@
+			ctx, cancel := context.WithTimeout(context.Background(), proposeTimeout)
+			defer cancel()
+			res, err := client.FetchCluster(ctx, &curpapi.FetchClusterRequest{})
+			if err != nil {
+				logrus.Debug(err)
+			} else {
+				fetchClusterRespCh <- res
+			}
+		}()
+	}
+
+	select {
+	case fetchClusterRes := <-fetchClusterRespCh:
+		return fetchClusterRes, nil
+	case <-time.After(proposeTimeout):
+		return nil, errors.New("fetch cluster fail")
+	}
+}
+
+// Propose the request to servers, if use_fast_path is false, it will wait for the synced index
+func (c protocolClient) Propose(cmd *xlineapi.Command, useFastPath bool) (*ProposeResponse, error) {
+	fastResCh := make(chan *ProposeResponse)
+	fastErrCh := make(chan error)
+	slowResCh := make(chan *ProposeResponse)
+	slowErrCh := make(chan error)
+
+	go func() {
+		proposeRes, err := c.slowRound(cmd)
+		slowResCh <- proposeRes
+		slowErrCh <- err
+	}()
+
+	go func() {
+		proposeRes, err := c.fastRound(cmd)
+		fastResCh <- proposeRes
+		fastErrCh <- err
+	}()
+
+	if useFastPath {
+		select {
+		case proposeRes := <-fastResCh:
+			err := <-fastErrCh
+			if proposeRes.IsFastRoundSuccess {
+				return proposeRes, err
+			}
+			proposeRes = <-slowResCh
+			err = <-slowErrCh
+			return proposeRes, err
+		case proposeRes := <-slowResCh:
+			err := <-slowErrCh
+			return proposeRes, err
+		}
+	} else {
+		<-fastResCh
+		<-fastErrCh
+		proposeRes := <-slowResCh
+		err := <-slowErrCh
+		return proposeRes, err
+	}
+}
+
+// The fast round of Curp protocol
+// It broadcast the requests to all the curp servers.
+func (c protocolClient) fastRound(cmd *xlineapi.Command) (*ProposeResponse, error) {
+	logrus.Info("Fast round start. Propose ID:", cmd.ProposeId)
+
+	isReceivedLeaderReq := false
+	isReceivedSuccessRes := 0
+	okCnt := 0
+	var clients []curpapi.ProtocolClient
+	var cmdRes xlineapi.CommandResponse
+	var proposeErr xlineapi.ExecuteError
+	proposeCh := make(chan *curpapi.ProposeResponse)
+
+	for _, m := range c.addrs {
+		conn, err := grpc.Dial(m, grpc.WithTransportCredentials(insecure.NewCredentials()), grpc.WithIdleTimeout(c.timeout.idleTimeout))
+		if err != nil {
+			logrus.Warn(err)
+		} else {
+			protocolClient := curpapi.NewProtocolClient(conn)
+			clients = append(clients[:], protocolClient)
+		}
+	}
+
+	serializeCmd, err := proto.Marshal(cmd)
+	if err != nil {
+		return nil, errors.New("command marshal fail")
+	}
+
+	for _, protocolClient := range clients {
+		client := protocolClient
+		go func() {
+			ctx, cancel := context.WithTimeout(context.Background(), c.timeout.proposeTimeout)
+			defer cancel()
+			res, err := client.Propose(ctx, &curpapi.ProposeRequest{Command: serializeCmd})
+			if err != nil {
+				logrus.Debug(err)
+				proposeCh <- nil
+			} else {
+				proposeCh <- res
+			}
+		}()
+	}
+
+	for i := 0; i < len(clients); i++ {
+		proposeRes := <-proposeCh
+
+		if proposeRes == nil {
+			continue
+		}
+
+		if proposeRes.ExeResult != nil {
+			isReceivedLeaderReq = true
+		}
```

Comment:
`ExeResult != nil` does not indicate that a response was received from the leader.

Reply:
done